### PR TITLE
Fix the "extract in finaliser" pattern.

### LINKED ISSUE #62: Finaliser runs too early
The follow program:
```
use "list.verona"

class A {
  _id: U64 & imm;

  create(id: U64 & imm): A & iso {
    var result = new A;
    result._id = id;
    result
  }

  final(self: mut) {
    Builtin.print1("Finalising A: {:#}\n", self)
  }
}

class Cell {
  _pool: cown[Pool] & imm;
  _f: A & iso;

  create(pool: cown[Pool] & imm, f: A & iso): Cell & iso {
    var result = new Cell;
    result._pool = pool;
    result._f = f;
    result
  }

  final(self: mut) {
    Builtin.print1("Finalising Cell: {:#}\n", self);
    Pool.return(self._pool, self._f = A.create(1));
  }
}

class Pool {
  _f: A & iso;

  create(): cown[Pool] & imm {
    var result = new Pool;
    cown.create(result)
  }

  return(pool: cown[Pool] & imm, f: A & iso) {
    Builtin.print1("Returning A: {:#}\n", mut-view f);
    when (var p = pool) {
      // FIXME: at this point A stops existing
      Builtin.print1("A: {:#}\n", mut-view f);
      p._f = f;
    }
  }
}

class Main {
  main() {
    var pool = Pool.create();
    var c = Cell.create(pool, A.create(0));
  }
}
```
When run gives the following output:
```
Finalising Cell: Cell { cown[Pool](0x7fde1b010020), A { 0 } }
Returning A: A { 0 }
Finalising A: A { 0 }
fish: '../verona/build/dist/veronac ./…' terminated by signal SIGSEGV (Address boundary error)
```
The value passed to `return` is finalised before it's passed into the behaviour.

### PR DESCRIPTION
**Pull request by plietar**
This adds a desctructor to every object. The destructor runs just before an object is deallocated. The VM uses this destructor to deallocate the array of fields, allowing us to move the "find iso fields" phase to after the finaliser has run.

Fixes #62.

### CODE REVIEW

**Comment by mjp41** [comment]
@mhyee could you take a look at this.  This is very close to the code from your internship, so this is probably very fresh for you.

**Comment by mjp41** [comment]
I don't see where this code calls destructors on arena objects that aren't large or have custom finalisers.

I think the choice is right for objects with destructors to end up in the plain ring for trace regions, but for arena regions it seems we actually want them in the finaliser end of the region, so objects without destructors or finalisers can be collected in bulk.  Otherwise, we will need to look at the whole region for dispose making it much more costly.

**Comment by mjp41** [comment]
I guess there is a design choice here.  If an object is complex in terms of its allocation, like the VM Objects with a second allocation for there fields, or when we add resizeable arrays, where again it will probably be two allocations.  Should the second allocation be in the arena?

Is it the case that "destructors" are only allowed to deallocate objects and nothing else?  Can they close file handles, or should that be the "finaliser"?

**Comment by jemc** [comment]
Could some documentation be included (for example, some code comments in `object.h` or elsewhere) that explains concretely the difference between a "finaliser" and a "destructor"; for example, to explain what each one is responsible for, when they run, and when to use one vs the other.

I know that there is some context on the need for this in issue #62, but it would be good if the repo itself had the documentation needed to understand the need for this, so that runtime API consumers can readily understand the fields they are responsible for populating in the descriptor without tracking it down in tickets outside of the code/docs.

File: src/rt/region/region_trace.h [hunk 1]
```diff
@@ -538,6 +541,14 @@ namespace verona::rt
       current_memory_used += size;
     }
 
+    static bool needs_finaliser_ring(Object* o)
+    {
+      // Objects with just a destructor don't need much special treatment, so
```
Comment (mhyee) [inline src/rt/region/region_trace.h]: Can you also update the comment at the top of this file?

File: src/rt/region/region_arena.h [hunk 2]
```diff
@@ -578,6 +580,21 @@ namespace verona::rt
       dealloc(alloc);
     }
 
+    /*
+     * Unlike traced regions, objects with a destructor but no finaliser are
+     * placed with the objects that do have a finaliser. This avoids having to
+     * walk all the objects on deallocation to find the ones with destructors.
+     */
+    static bool needs_finaliser_area(const Descriptor* desc)
```
Comment (mhyee) [inline src/rt/region/region_arena.h]: I don't think we ever agreed on what to call the finaliser and non-finaliser... areas... of an arena. My only nitpick with `area` is that it looks too much like `arena`. (`non-finaliser` is also a bit of a mouthful to say, and I guess it's no longer true, now that objects without finalisers but with destructors go into the other end.)

File: src/rt/region/region_arena.h [hunk 3]
```diff
@@ -578,6 +580,21 @@ namespace verona::rt
       dealloc(alloc);
     }
 
+    /*
+     * Unlike traced regions, objects with a destructor but no finaliser are
```
Comment (mhyee) [inline src/rt/region/region_arena.h]: Can you update the comment/diagram for the Arena class?

File: src/rt/region/region_arena.h (hunk 2 above)
Comment (mjp41) [reply src/rt/region/region_arena.h]: I think we had discusses the C++ non-trivially destructable.  But that is even more of a mouthful.

Rust uses the "Drop" term, so we could perhaps TrivialDrop and NonTrivialDrop.

File: src/interpreter/object.cc [hunk 4]
```diff
@@ -22,8 +22,16 @@ namespace verona::interpreter
   {
     rt::Descriptor::size = sizeof(VMObject);
     rt::Descriptor::trace = VMObject::trace_fn;
-    rt::Descriptor::trace_possibly_iso = VMObject::trace_fn;
-    rt::Descriptor::finaliser = VMObject::finaliser_fn;
+    rt::Descriptor::destructor = VMObject::destructor_fn;
```
Comment (mjp41) [inline src/interpreter/object.cc]: Could you use `is_trivially_destructible` trait here on VMObject to decide if we should have a destructor or not.  This will make the code more future proof.

File: src/interpreter/vm.cc [hunk 5]
```diff
@@ -67,30 +67,28 @@ namespace verona::interpreter
   void VM::execute_finaliser(VMObject* object)
   {
     const VMDescriptor* descriptor = object->descriptor();
-    uint32_t finaliser_ip = descriptor->finaliser_ip;
-    if (finaliser_ip != 0)
-    {
-      auto vm = VM::local_vm;
-      vm->trace("Finaliser for: {}", descriptor->name);
+    assert(descriptor->finaliser_ip > 0);
```
Comment (mjp41) [inline src/interpreter/vm.cc]: Nice change to make this code non-conditional by making the VM do the conditional work.

**Comment by plietar** [comment]
> Does src/rt/cpp/vobject.h need updating?

Currently vobject runs the C++ destructor in the verona finaliser. We could switch it to run as the verona destructor, and instead allow implementations to define a `finaliser` method.

**Comment by mjp41** [comment]
> > Does src/rt/cpp/vobject.h need updating?
> 
> Currently vobject runs the C++ destructor in the verona finaliser. We could switch it to run as the verona destructor, and instead allow implementations to define a `finaliser` method.

I think that makes sense from the runtime implementation.  As long as the C++ destructor only does C++ stuff, and doesn't do Verona stuff.  Previously, @davidchisnall and @theodus built a C++ wrapper classes to handle various forms of references, so then they performed Verona decrefs and region reclaims in destructors.  I think this semantics is not something we should support.

**Comment by plietar** [comment]
I’ve run into more difficulty than I expected, both in terms of edge cases and in trying to reduce the confusion in how rings/areas are named, so this might take me a bit longer to finish.

Meanwhile, I’ve found at least one existing C++ test that could benefit from this. The [Philosopher class](https://github.com/microsoft/verona/blob/master/src/rt/test/func/diningphilosophers/diningphil.cc#L58) has a vector of cowns, which currently gets deallocated in the finaliser.

There’s some tricks in place to work around this, but it should become simpler once we have proper destructors. It’s technically UB to call methods on the object after the C++ destructor has run, so a Sufficiently Smart Compiler could optimize the `!final` condition in the trace method (but neither clang nor gcc are that smart).

**Comment by plietar** [comment]
> As long as the C++ destructor only does C++ stuff, and doesn't do Verona stuff.

I think the finaliser must always leave the object in a “valid” state, as other finalisers running later could traverse the already finalised object and read its fields.

This means (if the RS isn’t used), decrementing reference counts must be done in the destructor, not the finaliser.

Obviously if you know the whole region topology you can play fast and loose (which the C++ tests do a lot), but in the Verona language we’ll have to stay safe.

**Comment by plietar** [comment]
I've finally gotten around to finishing this:
- The traced and arena regions use the same criteria to segregate objects.
- Objects are now categorized as trivial / non-trivial. An object is trivial if it has no finaliser, no destructor and no iso field.
- Traced regions need three iterations of the non-trivial ring. This is because all finalisers must run before we can start looking for sub-regions, and we must find all the sub-regions before we can start deallocating anything. The latter is because we look at the header bits of the object pointed by the field to decide whether it is a child region. If we had enough type information from the compiler, we could maybe merge the last two iterations.
- The cown leak detector now looks at every object, not just the ones on the non-trivial ring. This allows fewer objects to appear on that ring, optimizing the common non-LD code paths.
